### DISCUSSION TITLE
Are non-shared virtual environments automatically activated?

### DISCUSSION THREAD

**Discussion by sueskind** [lucasvieirasilva/nx-plugins#179] [Q&A] [answered]
I would like to add the following target to my `project.json`:

```json
"behave": {
    "executor": "@nxlv/python:run-commands",
    "options": {
        "command": "behave",
        "forwardAllArgs": true,
        "cwd": "apps/my-project"
    }
}
```

It runs the [`behave`](https://github.com/behave/behave) command. Of course, for this to succeed, the venv must be activated.

I found a lot of discussions regarding venv activation for the `@nxlv/python:run-commands` executor in #128, #132, the [blog post](https://betterprogramming.pub/poetry-python-nx-monorepo-5750d8627024) and [this section](https://github.com/lucasvieirasilva/nx-plugins/tree/main/packages/nx-python#run-commands-same-as-nxrun-commands) in the readme. However, they all discuss the case where a shared environment is used.

I use a separate venv for the project and when running the target with `nx run my-project:behave`, I get the error `/bin/sh: line 1: behave: command not found`. When I activate the venv manually beforehand, the command runs without issues.

So my question is: Is the venv supposed to be activated automatically for non-shared virtual environments?

- If yes, what may I be doing wrong?
- If no, what is the intended/recommended way of handling the venv activation for custom targets?

**Comment by lucasvieirasilva** [comment, marked as answer]
Hi @sueskind, currently, only shared environments are automatically activated because in the cases you mentioned you can use the `poetry run ...` before your command and poetry takes care of running your command in the virtual environment.

Example:

```json
"behave": {
    "executor": "@nxlv/python:run-commands",
    "options": {
        "command": "poetry run behave",
        "forwardAllArgs": true,
        "cwd": "apps/my-project"
    }
}
```

**Reply by sueskind** [reply]
Thank you very much for the quick response!

Ah, I see. I am not experienced with poetry (I'm mostly using "raw" venvs), so I did not know that you run commands via `poetry run`. :innocent:
